File: unrealmcp/tools/materials.py
"""Material tools — create, inspect, and build complex material graphs.

All operations go through C++ FEpicUnrealMCPMaterialCommands via the TCP bridge.
"""

import json
from typing import Optional

from unrealmcp._bridge import mcp
from unrealmcp._tcp_bridge import _call


@mcp.tool()
def create_material(
    name: str,
    path: str = "/Game/Materials",
    blend_mode: str = "opaque",
    shading_model: str = "default_lit",
    two_sided: bool = False,
    opacity_mask_clip_value: Optional[float] = None,
) -> str:
    """Create a new Material asset.

    Args:
        name: Material name (e.g. "M_MyMaterial")
        path: Content Browser path
        blend_mode: opaque | masked | translucent | additive | modulate | alpha_composite | alpha_holdout
        shading_model: default_lit | unlit | subsurface | clear_coat | subsurface_profile |
            two_sided_foliage | cloth | eye | thin_translucent
        two_sided: Render on both sides
        opacity_mask_clip_value: Clip value for masked blend mode (0.0-1.0)
    """
    params = {
        "name": name,
        "path": path,
        "blend_mode": blend_mode,
        "shading_model": shading_model,
        "two_sided": two_sided,
    }
    if opacity_mask_clip_value is not None:
        params["opacity_mask_clip_value"] = opacity_mask_clip_value
    return _call("create_material", params)


@mcp.tool()
def create_material_instance(
    parent_path: str,
    name: str,
    path: str = "/Game/Materials",
    scalar_params: Optional[str] = None,
    vector_params: Optional[str] = None,
    texture_params: Optional[str] = None,
) -> str:
    """Create a Material Instance from a parent material.

    Args:
        parent_path: Full path to parent (e.g. "/Game/Materials/M_Base")
        name: Instance name (e.g. "MI_Red")
        scalar_params: JSON dict of scalar overrides — {"Opacity": 0.5, "Metallic": 1.0}
        vector_params: JSON dict of vector overrides — {"BaseColor": [1.0, 0.0, 0.0, 1.0]}
        texture_params: JSON dict of texture overrides — {"Texture": "/Game/Textures/T_Wood"}
    """
    params = {
        "parent_path": parent_path,
        "name": name,
        "path": path,
    }
    if scalar_params is not None:
        params["scalar_params"] = json.loads(scalar_params)
    if vector_params is not None:
        params["vector_params"] = json.loads(vector_params)
    if texture_params is not None:
        params["texture_params"] = json.loads(texture_params)
    return _call("create_material_instance", params)


@mcp.tool()
def build_material_graph(
    material_path: str,
    nodes: str,
    connections: str,
    clear_existing: bool = True,
) -> str:
    """Build a complete material node graph in one atomic operation.

    Creates expression nodes and wires them together in a single call.
    By default clears existing expressions first (safe rebuild — all external
    references stay intact).

    Args:
        material_path: Full path to existing material (e.g. "/Game/Materials/M_Portal")
        nodes: JSON array of node definitions. Each node has:
            - type: Short class name (e.g. "TextureCoordinate", "Custom", "ScalarParameter",
                "Constant3Vector", "TextureSample", "Multiply", "Add", "Panner", "Time",
                "VectorParameter", "LinearInterpolate", "ComponentMask")
                Auto-prefixed with "MaterialExpression" if needed.
            - pos_x, pos_y: Graph position (default -300, 0)
            - properties: Dict of editor properties to set. Common ones:
                - parameter_name: str — name shown in material instances
                - default_value: number or [r,g,b,a]
                - slider_min, slider_max: float — slider range
                - group: str — parameter group name
                - Values starting with "/" are loaded as assets
                - Lists of 3-4 become LinearColor, lists of 2 become Vector2D
            For Custom HLSL nodes (type="Custom"), also supports:
            - code: HLSL source code string
            - description: Node title in the graph
            - output_type: "float" | "float2" | "float3" | "float4"
            - inputs: List of input pin names (e.g. ["UV", "Speed"])
            - outputs: List of {"name": str, "type": str} for additional outputs
        connections: JSON array of connections. Each connection has:
            - from_node: Source node index (int, 0-based into nodes array)
            - from_pin: Output pin name ("" for default output)
            - to_node: Target node index (int) or "material" for material output
            - to_pin: Input pin name, or material property when to_node="material"
                (BaseColor, Metallic, Roughness, EmissiveColor, Opacity, OpacityMask,
                Normal, Specular, AmbientOcclusion, WorldPositionOffset, SubsurfaceColor,
                Refraction)
        clear_existing: Remove existing expressions before building (default True)
    """
    return _call("build_material_graph", {
        "material_path": material_path,
        "nodes": json.loads(nodes),
        "connections": json.loads(connections),
        "clear_existing": clear_existing,
    })


@mcp.tool()
def get_material_info(
    material_path: str,
    include: Optional[str] = None,
) -> str:
    """Inspect a material's properties.

    Always returns: blend mode, shading model, two-sided, expression count.

    Use include (comma-separated) to request additional sections:
    - "parameters": scalar/vector/texture parameter names
    - "textures": used texture paths
    - "statistics": shader instruction counts

    Omit include to get only basic properties (minimal tokens).
    Pass include="parameters,textures,statistics" for everything.
    """
    params: dict = {"material_path": material_path}
    if include is not None:
        params["include"] = include
    return _call("get_material_info", params)


@mcp.tool()
def recompile_material(material_path: str) -> str:
    """Force recompile and save a material. Reports success/failure."""
    return _call("recompile_material", {"material_path": material_path})


@mcp.tool()
def get_material_errors(material_path: str, recompile: bool = True) -> str:
    """Get shader compilation errors for a material.

    Returns error messages and the node indices that caused them.
    Recompiles first by default to get fresh errors.
    """
    return _call("get_material_errors", {
        "material_path": material_path,
        "recompile": recompile,
    })


@mcp.tool()
def set_material_properties(
    material_path: str,
    blend_mode: Optional[str] = None,
    shading_model: Optional[str] = None,
    two_sided: Optional[bool] = None,
    opacity_mask_clip_value: Optional[float] = None,
    dithered_lof_transition: Optional[bool] = None,
    allow_negative_emissive_color: Optional[bool] = None,
    recompile: bool = True,
) -> str:
    """Bulk-set material-level properties in one call.

    Args:
        blend_mode: opaque | masked | translucent | additive | modulate | alpha_composite | alpha_holdout
        shading_model: default_lit | unlit | subsurface | clear_coat | etc.
        recompile: Recompile after setting properties (default True)
    """
    params: dict = {
        "material_path": material_path,
        "recompile": recompile,
    }
    for key, val in [
        ("blend_mode", blend_mode), ("shading_model", shading_model),
        ("two_sided", two_sided), ("opacity_mask_clip_value", opacity_mask_clip_value),
        ("dithered_lof_transition", dithered_lof_transition),
        ("allow_negative_emissive_color", allow_negative_emissive_color),
    ]:
        if val is not None:
            params[key] = val
    return _call("set_material_properties", params)


@mcp.tool()
def get_material_graph_nodes(
    material_path: str,
    verbosity: str = "connections",
    type_filter: Optional[str] = None,
) -> str:
    """Read expression nodes in a material graph.

    Use verbosity to control response size:
    - "summary": index, type, position only (~30 tokens/node)
    - "connections": + input connections (default, good balance)
    - "full": + properties + available pins (large response)

    Use type_filter to only return nodes matching a type substring
    (e.g. "Parameter" returns only parameter nodes).
    """
    params: dict = {
        "material_path": material_path,
        "verbosity": verbosity,
    }
    if type_filter is not None:
        params["type_filter"] = type_filter
    return _call("get_material_graph_nodes", params)


@mcp.tool()
def add_material_expression(material_path: str, node: str) -> str:
    """Add a single expression node without clearing other nodes.

    Returns the new node_index. Wire it up with connect_material_expressions.

    node is a JSON object: {"type": "ScalarParameter", "pos_x": -1200, "pos_y": -400,
    "properties": {"parameter_name": "Speed", "default_value": 0.5}}

    For Custom HLSL nodes, include top-level "code", "description", "output_type",
    "inputs", and "outputs" fields.
    """
    return _call("add_material_expression", {
        "material_path": material_path,
        "node": json.loads(node),
    })


@mcp.tool()
def connect_material_expressions(
    material_path: str,
    from_node: int,
    to_node: str,
    to_pin: str,
    from_pin: str = "",
) -> str:
    """Connect two expression nodes using their indices.

    Args:
        from_node: Source node index
        to_node: Target node index (e.g. "5") or "material" for material output
        to_pin: Input pin name, or material property when to_node="material"
            (BaseColor, Metallic, Roughness, EmissiveColor, Opacity, etc.)
        from_pin: Output pin name ("" = primary output)
    """
    return _call("connect_material_expressions", {
        "material_path": material_path,
        "from_node": from_node,
        "from_pin": from_pin,
        "to_node": to_node,
        "to_pin": to_pin,
    })


@mcp.tool()
def delete_material_expression(material_path: str, node_index: int) -> str:
    """Delete a single expression node by index. Recompiles and saves automatically.

    Remaining node indices may shift — re-query with get_material_graph_nodes afterwards.
    """
    return _call("delete_material_expression", {
        "material_path": material_path,
        "node_index": node_index,
    })


@mcp.tool()
def add_material_comments(material_path: str, comments: str) -> str:
    """Add comment boxes to a material graph for organization.

    comments is a JSON array. Each entry: {text, pos_x, pos_y, size_x, size_y,
    font_size, color: [r,g,b], show_bubble, color_bubble, group_mode}.
    """
    return _call("add_material_comments", {
        "material_path": material_path,
        "comments": json.loads(comments),
    })


@mcp.tool()
def get_material_expression_info(material_path: str, node_index: int) -> str:
    """Get detailed info for a single node, including all available input/output pins.

    Use this before connecting nodes to discover exact pin names.
    """
    return _call("get_material_expression_info", {
        "material_path": material_path,
        "node_index": node_index,
    })


@mcp.tool()
def get_material_property_connections(material_path: str) -> str:
    """Query which expression node feeds each material output slot.

    Only lists slots that have something connected.
    """
    return _call("get_material_property_connections", {"material_path": material_path})


@mcp.tool()
def set_material_expression_property(
    material_path: str,
    node_index: int,
    property_name: str,
    property_value: str,
) -> str:
    """Set a property on an existing material expression node.

    Accepts snake_case ("parameter_name") or PascalCase ("ParameterName").

    Common properties: parameter_name, default_value, slider_min, slider_max,
    group, texture, code, description, output_type, inputs, add_inputs, outputs,
    SpeedX, SpeedY, CoordinateIndex, R, Constant.
    """
    return _call("set_material_expression_property", {
        "material_path": material_path,
        "node_index": node_index,
        "property_name": property_name,
        "property_value": json.loads(property_value),
    })


@mcp.tool()
def move_material_expression(material_path: str, node_index: int, pos_x: int, pos_y: int) -> str:
    """Move a material expression node to a new graph position."""
    return _call("move_material_expression", {
        "material_path": material_path,
        "node_index": node_index,
        "pos_x": pos_x,
        "pos_y": pos_y,
    })


@mcp.tool()
def duplicate_material_expression(
    material_path: str,
    node_index: int,
    offset_x: int = 0,
    offset_y: int = 150,
) -> str:
    """Duplicate a node (same type and properties, offset from original).

    Returns the new node_index. Connections are NOT copied.
    """
    return _call("duplicate_material_expression", {
        "material_path": material_path,
        "node_index": node_index,
        "offset_x": offset_x,
        "offset_y": offset_y,
    })


@mcp.tool()
def layout_material_expressions(material_path: str) -> str:
    """Auto-layout all nodes using Unreal's built-in arrangement algorithm."""
    return _call("layout_material_expressions", {"material_path": material_path})


@mcp.tool()
def get_material_instance_parameters(material_path: str) -> str:
    """Get all overridable parameters from a Material Instance (scalar, vector, texture, switch)."""
    return _call("get_material_instance_parameters", {"material_path": material_path})


@mcp.tool()
def set_material_instance_parameter(
    material_path: str,
    param_name: str,
    param_type: str,
    value: str,
) -> str:
    """Set a parameter override on a Material Instance.

    Args:
        param_type: scalar | vector | texture | static_switch
        value: JSON-encoded — float, [r,g,b,a], "/Game/path", or true/false
    """
    return _call("set_material_instance_parameter", {
        "material_path": material_path,
        "param_name": param_name,
        "param_type": param_type,
        "value": json.loads(value),
    })


@mcp.tool()
def list_material_expression_types(
    filter: Optional[str] = None,
    max_results: int = 0,
    include_details: bool = True,
) -> str:
    """List available material expression node types.

    Use filter to narrow results (e.g. "texture", "parameter", "math", "noise").
    Set max_results to limit output (0 = unlimited).
    Set include_details=false for compact output (type names only).
    """
    params: dict = {}
    if filter is not None:
        params["filter"] = filter
    if max_results > 0:
        params["max_results"] = max_results
    if not include_details:
        params["include_details"] = False
    return _call("list_material_expression_types", params)


@mcp.tool()
def get_expression_type_info(type_name: str) -> str:
    """Look up pin names and editable properties for a node type WITHOUT creating one.

    Returns input pins, output pins, and all editable properties with types and defaults.
    Use this BEFORE creating nodes to know exact pin names and avoid connection errors.

    Args:
        type_name: Short type name (e.g. "Multiply", "TextureSample", "Noise", "Custom")
    """
    return _call("get_expression_type_info", {"type_name": type_name})


@mcp.tool()
def disconnect_material_expression(
    material_path: str,
    node_index: int,
    input_pin: str,
) -> str:
    """Disconnect a specific input pin on a material expression node.

    Args:
        node_index: The target node whose input pin will be disconnected
        input_pin: Name of the input pin to disconnect
    """
    return _call("disconnect_material_expression", {
        "material_path": material_path,
        "node_index": node_index,
        "input_pin": input_pin,
    })


@mcp.tool()
def search_material_functions(
    filter: Optional[str] = None,
    path: str = "/Game",
    max_results: int = 50,
    include_engine: bool = False,
) -> str:
    """Search for Material Functions by name.

    Args:
        filter: Name substring to match (e.g. "Blend", "Normal", "Fresnel")
        path: Content Browser path to search in (default "/Game")
        max_results: Maximum results to return (default 50)
        include_engine: Include engine built-in material functions
    """
    params: dict = {
        "path": path,
        "max_results": max_results,
        "include_engine": include_engine,
    }
    if filter is not None:
        params["filter"] = filter
    return _call("search_material_functions", params)


@mcp.tool()
def validate_material_graph(material_path: str) -> str:
    """Diagnose connection issues in a material graph.

    Categorises every node into:
    - orphaned: no inputs AND output not consumed — safe to delete
    - dead_ends: has inputs but output goes nowhere
    - missing_inputs: some input pins are empty
    - unconnected_inputs: ALL input pins are empty (node expects connections)
    - unconnected_outputs: output not consumed by anything

    Returns healthy=true when no orphaned or dead-end nodes exist.
    """
    return _call("validate_material_graph", {"material_path": material_path})


@mcp.tool()
def trace_material_connection(
    material_path: str,
    node_index: int,
    direction: str = "both",
    max_depth: int = 1,
) -> str:
    """Trace connections upstream and/or downstream from a specific node.

    Shows exactly what feeds into each input pin and what consumes each output pin,
    including connections to material outputs (BaseColor, Normal, etc.).

    Args:
        node_index: The node to trace from
        direction: "upstream" (what feeds in), "downstream" (what consumes output),
            or "both" (default)
        max_depth: How many hops to trace recursively (1 = immediate neighbors,
            2+ = follow the chain). Max 50. Default 1.
    """
    return _call("trace_material_connection", {
        "material_path": material_path,
        "node_index": node_index,
        "direction": direction,
        "max_depth": max_depth,
    })


@mcp.tool()
def cleanup_material_graph(
    material_path: str,
    mode: str = "orphaned",
    dry_run: bool = False,
) -> str:
    """Delete orphaned and/or dead-end nodes from a material graph.

    WARNING: NEVER call this tool unless the user explicitly asks to clean up
    or delete unconnected nodes. Users may intentionally keep disconnected nodes
    as references, scratchpads, or work-in-progress. Deleting them without
    permission will cause data loss and frustration.

    Args:
        mode: What to delete:
            "orphaned" — only nodes with NO connections at all (safest)
            "dead_ends" — nodes whose output goes nowhere (has inputs but no consumers)
            "all" — both orphaned and dead-ends
        dry_run: If true, reports what WOULD be deleted without actually deleting.
            Always use dry_run=true first to preview before deleting.
    """
    return _call("cleanup_material_graph", {
        "material_path": material_path,
        "mode": mode,
        "dry_run": dry_run,
    })

File: cli/cmd/materials.go
package cmd

func init() {
	ensureGroup("materials", "Materials")
	registerCommands(materialCommands)
}

var materialCommands = []CommandSpec{
	{
		Name:  "create_material",
		Group: "materials",
		Short: "Create a new Material",
		Params: []ParamSpec{
			{Name: "name", Type: "string", Required: true, Help: "Material name"},
			{Name: "path", Type: "string", Default: "/Game/Materials", Help: "Content path"},
			{Name: "blend_mode", Type: "string", Default: "opaque", Help: "Blend mode"},
			{Name: "shading_model", Type: "string", Default: "default_lit", Help: "Shading model"},
			{Name: "two_sided", Type: "bool", Default: false, Help: "Two-sided rendering"},
			{Name: "opacity_mask_clip_value", Type: "float", Help: "Opacity mask clip value"},
		},
	},
	{
		Name:  "create_material_instance",
		Group: "materials",
		Short: "Create a Material Instance",
		Params: []ParamSpec{
			{Name: "parent_path", Type: "string", Required: true, Help: "Parent material path"},
			{Name: "name", Type: "string", Required: true, Help: "Instance name"},
			{Name: "path", Type: "string", Default: "/Game/Materials", Help: "Content path"},
			{Name: "scalar_params", Type: "json", Help: "JSON string of scalar parameters"},
			{Name: "vector_params", Type: "json", Help: "JSON string of vector parameters"},
			{Name: "texture_params", Type: "json", Help: "JSON string of texture parameters"},
		},
	},
	{
		Name:    "build_material_graph",
		Group:   "materials",
		Short:   "Build complete node graph in one atomic call",
		LargeOp: true,
		Params: []ParamSpec{
			{Name: "material_path", Type: "string", Required: true, Help: "Material asset path"},
			{Name: "nodes", Type: "json", Required: true, Help: "JSON array of node definitions"},
			{Name: "connections", Type: "json", Required: true, Help: "JSON array of connections"},
			{Name: "clear_existing", Type: "bool", Default: true, Help: "Clear existing expressions"},
		},
	},
	{
		Name:  "get_material_info",
		Group: "materials",
		Short: "Inspect material properties, parameters, textures",
		Params: []ParamSpec{
			{Name: "material_path", Type: "string", Required: true, Help: "Material asset path"},
			{Name: "include", Type: "string", Help: "Comma-sep: parameters,textures,statistics"},
		},
	},
	{
		Name:  "set_material_properties",
		Group: "materials",
		Short: "Bulk-set material properties",
		Params: []ParamSpec{
			{Name: "material_path", Type: "string", Required: true, Help: "Material asset path"},
			{Name: "blend_mode", Type: "string", Help: "Blend mode"},
			{Name: "shading_model", Type: "string", Help: "Shading model"},
			{Name: "two_sided", Type: "bool", Help: "Two-sided"},
			{Name: "opacity_mask_clip_value", Type: "float", Help: "Clip value"},
			{Name: "dithered_lof_transition", Type: "bool", Help: "Dithered LOF transition"},
			{Name: "allow_negative_emissive_color", Type: "bool", Help: "Allow negative emissive"},
			{Name: "recompile", Type: "bool", Default: true, Help: "Recompile after changes"},
		},
	},
	{
		Name:  "recompile_material",
		Group: "materials",
		Short: "Force recompile and save",
		Params: []ParamSpec{
			{Name: "material_path", Type: "string", Required: true, Help: "Material asset path"},
		},
	},
	{
		Name:  "get_material_errors",
		Group: "materials",
		Short: "Get material compilation errors",
		Params: []ParamSpec{
			{Name: "material_path", Type: "string", Required: true, Help: "Material asset path"},
			{Name: "recompile", Type: "bool", Default: true, Help: "Recompile first"},
		},
	},
	{
		Name:  "add_material_comments",
		Group: "materials",
		Short: "Add comment boxes to material graph",
		Params: []ParamSpec{
			{Name: "material_path", Type: "string", Required: true, Help: "Material asset path"},
			{Name: "comments", Type: "json", Required: true, Help: "JSON array of comment definitions"},
		},
	},
	{
		Name:  "get_material_graph_nodes",
		Group: "materials",
		Short: "Read graph nodes with verbosity control",
		Params: []ParamSpec{
			{Name: "material_path", Type: "string", Required: true, Help: "Material asset path"},
			{Name: "type_filter", Type: "string", Help: "Filter by node type"},
			{Name: "verbosity", Type: "string", Default: "connections", Help: "summary, connections, or full"},
		},
	},
	{
		Name:  "add_material_expression",
		Group: "materials",
		Short: "Add a material expression node",
		Params: []ParamSpec{
			{Name: "material_path", Type: "string", Required: true, Help: "Material asset path"},
			{Name: "node", Type: "json", Required: true, Help: "JSON node definition"},
		},
	},
	{
		Name:  "connect_material_expressions",
		Group: "materials",
		Short: "Connect two material expression nodes",
		Params: []ParamSpec{
			{Name: "material_path", Type: "string", Required: true, Help: "Material asset path"},
			{Name: "from_node", Type: "int", Required: true, Help: "Source node index"},
			{Name: "to_node", Type: "string", Required: true, Help: "Target node or 'material'"},
			{Name: "to_pin", Type: "string", Required: true, Help: "Target pin name"},
			{Name: "from_pin", Type: "string", Help: "Source pin name"},
		},
	},
	{
		Name:  "delete_material_expression",
		Group: "materials",
		Short: "Delete a material expression by index",
		Params: []ParamSpec{
			{Name: "material_path", Type: "string", Required: true, Help: "Material asset path"},
			{Name: "node_index", Type: "int", Required: true, Help: "Node index"},
		},
	},
	{
		Name:  "set_material_expression_property",
		Group: "materials",
		Short: "Set a property on a material expression",
		Params: []ParamSpec{
			{Name: "material_path", Type: "string", Required: true, Help: "Material asset path"},
			{Name: "node_index", Type: "int", Required: true, Help: "Node index"},
			{Name: "property_name", Type: "string", Required: true, Help: "Property name"},
			{Name: "property_value", Type: "string", Required: true, Help: "Property value"},
		},
	},
	{
		Name:  "move_material_expression",
		Group: "materials",
		Short: "Move a material expression node",
		Params: []ParamSpec{
			{Name: "material_path", Type: "string", Required: true, Help: "Material asset path"},
			{Name: "node_index", Type: "int", Required: true, Help: "Node index"},
			{Name: "pos_x", Type: "int", Required: true, Help: "X position"},
			{Name: "pos_y", Type: "int", Required: true, Help: "Y position"},
		},
	},
	{
		Name:  "duplicate_material_expression",
		Group: "materials",
		Short: "Duplicate a material expression",
		Params: []ParamSpec{
			{Name: "material_path", Type: "string", Required: true, Help: "Material asset path"},
			{Name: "node_index", Type: "int", Required: true, Help: "Node index to duplicate"},
			{Name: "offset_x", Type: "int", Default: 0, Help: "X offset"},
			{Name: "offset_y", Type: "int", Default: 150, Help: "Y offset"},
		},
	},
	{
		Name:  "layout_material_expressions",
		Group: "materials",
		Short: "Auto-layout all material expressions",
		Params: []ParamSpec{
			{Name: "material_path", Type: "string", Required: true, Help: "Material asset path"},
		},
	},
	{
		Name:  "get_material_instance_parameters",
		Group: "materials",
		Short: "Get all parameters of a material instance",
		Params: []ParamSpec{
			{Name: "material_path", Type: "string", Required: true, Help: "Material instance path"},
		},
	},
	{
		Name:  "set_material_instance_parameter",
		Group: "materials",
		Short: "Set a parameter on a material instance",
		Params: []ParamSpec{
			{Name: "material_path", Type: "string", Required: true, Help: "Material instance path"},
			{Name: "param_name", Type: "string", Required: true, Help: "Parameter name"},
			{Name: "param_type", Type: "string", Required: true, Help: "Parameter type"},
			{Name: "value", Type: "string", Required: true, Help: "Parameter value"},
		},
	},
	{
		Name:  "list_material_expression_types",
		Group: "materials",
		Short: "List available material expression types",
		Params: []ParamSpec{
			{Name: "filter", Type: "string", Help: "Filter by type name"},
			{Name: "max_results", Type: "int", Default: 0, Help: "Max results (0 = all)"},
			{Name: "include_details", Type: "bool", Default: true, Help: "Include type details"},
		},
	},
	{
		Name:  "get_expression_type_info",
		Group: "materials",
		Short: "Look up pins & properties for a node type",
		Params: []ParamSpec{
			{Name: "type_name", Type: "string", Required: true, Help: "Expression type (e.g. Multiply, TextureSample)"},
		},
	},
	{
		Name:  "disconnect_material_expression",
		Group: "materials",
		Short: "Disconnect a specific input pin",
		Params: []ParamSpec{
			{Name: "material_path", Type: "string", Required: true, Help: "Material asset path"},
			{Name: "node_index", Type: "int", Required: true, Help: "Node index"},
			{Name: "input_pin", Type: "string", Required: true, Help: "Input pin name"},
		},
	},
	{
		Name:  "search_material_functions",
		Group: "materials",
		Short: "Find Material Functions by name",
		Params: []ParamSpec{
			{Name: "filter", Type: "string", Help: "Name filter"},
			{Name: "path", Type: "string", Default: "/Game", Help: "Search path"},
			{Name: "max_results", Type: "int", Default: 50, Help: "Max results"},
			{Name: "include_engine", Type: "bool", Default: false, Help: "Include engine functions"},
		},
	},
	{
		Name:  "validate_material_graph",
		Group: "materials",
		Short: "Diagnose orphaned, dead-end, and unconnected nodes",
		Params: []ParamSpec{
			{Name: "material_path", Type: "string", Required: true, Help: "Material asset path"},
		},
	},
	{
		Name:  "trace_material_connection",
		Group: "materials",
		Short: "Trace upstream/downstream from a node",
		Params: []ParamSpec{
			{Name: "material_path", Type: "string", Required: true, Help: "Material asset path"},
			{Name: "node_index", Type: "int", Required: true, Help: "Node index"},
			{Name: "direction", Type: "string", Default: "both", Help: "upstream, downstream, or both"},
			{Name: "max_depth", Type: "int", Default: 1, Help: "Max trace depth"},
		},
	},
	{
		Name:  "cleanup_material_graph",
		Group: "materials",
		Short: "Delete orphaned/dead-end nodes",
		Params: []ParamSpec{
			{Name: "material_path", Type: "string", Required: true, Help: "Material asset path"},
			{Name: "mode", Type: "string", Default: "orphaned", Help: "orphaned, dead_ends, or all"},
			{Name: "dry_run", Type: "bool", Default: false, Help: "Preview without deleting"},
		},
	},

	// -- Material Functions --
	{
		Name:  "create_material_function",
		Group: "materials",
		Short: "Create a new Material Function asset",
		Params: []ParamSpec{
			{Name: "name", Type: "string", Required: true, Help: "Function name"},
			{Name: "path", Type: "string", Default: "/Game/Materials/Functions", Help: "Content path"},
			{Name: "description", Type: "string", Help: "Description"},
			{Name: "expose_to_library", Type: "bool", Default: true, Help: "Expose to material library"},
		},
	},
	{
		Name:  "get_material_function_info",
		Group: "materials",
		Short: "Inspect a Material Function",
		Params: []ParamSpec{
			{Name: "function_path", Type: "string", Required: true, Help: "Function asset path"},
		},
	},
	{
		Name:    "build_material_function_graph",
		Group:   "materials",
		Short:   "Build complete MF node graph atomically",
		LargeOp: true,
		Params: []ParamSpec{
			{Name: "function_path", Type: "string", Required: true, Help: "Function asset path"},
			{Name: "nodes", Type: "json", Required: true, Help: "JSON array of node definitions"},
			{Name: "connections", Type: "json", Required: true, Help: "JSON array of connections"},
			{Name: "clear_existing", Type: "bool", Default: true, Help: "Clear existing expressions"},
		},
	},
	{
		Name:  "add_material_function_input",
		Group: "materials",
		Short: "Add a FunctionInput pin",
		Params: []ParamSpec{
			{Name: "function_path", Type: "string", Required: true, Help: "Function asset path"},
			{Name: "input_name", Type: "string", Required: true, Help: "Input name"},
			{Name: "input_type", Type: "string", Default: "Scalar", Help: "Input type"},
			{Name: "description", Type: "string", Help: "Description"},
			{Name: "sort_priority", Type: "int", Default: 0, Help: "Sort priority"},
			{Name: "use_preview_as_default", Type: "bool", Default: false, Help: "Use preview as default"},
			{Name: "preview_value", Type: "string", Help: "Preview value"},
			{Name: "pos_x", Type: "int", Default: -600, Help: "X position"},
			{Name: "pos_y", Type: "int", Default: 0, Help: "Y position"},
		},
	},
	{
		Name:  "add_material_function_output",
		Group: "materials",
		Short: "Add a FunctionOutput pin",
		Params: []ParamSpec{
			{Name: "function_path", Type: "string", Required: true, Help: "Function asset path"},
			{Name: "output_name", Type: "string", Required: true, Help: "Output name"},
			{Name: "description", Type: "string", Help: "Description"},
			{Name: "sort_priority", Type: "int", Default: 0, Help: "Sort priority"},
			{Name: "pos_x", Type: "int", Default: 200, Help: "X position"},
			{Name: "pos_y", Type: "int", Default: 0, Help: "Y position"},
		},
	},
	{
		Name:  "set_material_function_input",
		Group: "materials",
		Short: "Modify an existing FunctionInput",
		Params: []ParamSpec{
			{Name: "function_path", Type: "string", Required: true, Help: "Function asset path"},
			{Name: "node_index", Type: "int", Required: true, Help: "Node index"},
			{Name: "input_name", Type: "string", Help: "Input name"},
			{Name: "input_type", Type: "string", Help: "Input type"},
			{Name: "description", Type: "string", Help: "Description"},
			{Name: "sort_priority", Type: "int", Help: "Sort priority"},
			{Name: "use_preview_as_default", Type: "bool", Help: "Use preview as default"},
			{Name: "preview_value", Type: "string", Help: "Preview value"},
		},
	},
	{
		Name:  "set_material_function_output",
		Group: "materials",
		Short: "Modify an existing FunctionOutput",
		Params: []ParamSpec{
			{Name: "function_path", Type: "string", Required: true, Help: "Function asset path"},
			{Name: "node_index", Type: "int", Required: true, Help: "Node index"},
			{Name: "output_name", Type: "string", Help: "Output name"},
			{Name: "description", Type: "string", Help: "Description"},
			{Name: "sort_priority", Type: "int", Help: "Sort priority"},
		},
	},
	{
		Name:  "validate_material_function",
		Group: "materials",
		Short: "Validate a Material Function",
		Params: []ParamSpec{
			{Name: "function_path", Type: "string", Required: true, Help: "Function asset path"},
		},
	},
	{
		Name:  "cleanup_material_function",
		Group: "materials",
		Short: "Cleanup orphaned nodes in a Material Function",
		Params: []ParamSpec{
			{Name: "function_path", Type: "string", Required: true, Help: "Function asset path"},
			{Name: "dry_run", Type: "bool", Default: false, Help: "Preview without deleting"},
		},
	},
}
